lib/galaxy/celery/tasks.py

```diff
@@ -98,15 +99,61 @@ def set_job_metadata(
     )
 
 
+@galaxy_task(action="set or detect dataset datatype")
+def change_datatype(
+    hda_manager: HDAManager,
+    ldda_manager: LDDAManager,
+    datatypes_registry: DatatypesRegistry,
+    sa_session: galaxy_scoped_session,
+    dataset_id: int,
+    datatype: str,
+    model_class: str = "HistoryDatasetAssociation",
+):
+    dataset_instance = _get_dataset_by_id(hda_manager, ldda_manager, dataset_id, model_class)
+    if datatype == "auto":
+        path = dataset_instance.dataset.file_name
+        datatype = sniff.guess_ext(path, datatypes_registry.sniff_order)
+    datatypes_registry.change_datatype(dataset_instance, datatype)
+    sa_session.flush()
```

Comment:
I'd call directly into the set_metadata task here. If you call it without `delay` it runs like your normal function.

Reply:
Ohh, was just trying to chain both tasks but took me some time until I figured out the chaining was messing with the dependency injection and need to use `immutable signatures` :sweat_smile: 

I think it works now, do you still think is better to call it directly in set_metadata?

Reply:
Yes, those don't need to be separate tasks, they have significant overhead and it's really one atomic operation.

Reply:
Makes total sense, done!